### PR TITLE
Support fetching docker images from V2 registries

### PR DESCRIPTION
**Pull request by ixmatus**
###### Motivation for this change
The stock `dockerTools` in `nixpkgs` support fetching docker images from a V1 docker registry and not a V2 registry; `dockerTools` pull the whole docker image into the nix store instead of pulling the layers, config, and manifest into the nix store. This is problematic because `dockerTools` only work with an out-of-date docker registry version and it is storage-inefficient; inefficient because multiple docker images sharing base layers do not take advantage of deduplication in the nix store.

[`hocker`](https://github.com/awakesecurity/hocker#readme) and the `fetchdocker` utilities fix these issues. `hocker` implements its own client against the docker distribution REST API and it does not use the docker client or docker daemon to pull an image from a registry. Each utility of `hocker` is designed to do one thing well, this translates naturally to a set of Nix functions that produce derivations that fetch docker image layers, the docker image configuration json, and produce an image compositor script. The image compositor script assembles the downloaded and the generated pieces of the docker image from the nix store, streaming it to stdout as a `tar` archive for use with `docker load`.

###### Things TODO next, after merging
- [ ] Write a tutorial
- [ ] Contribute a manual entry

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [X] macOS
   - [X] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by ixmatus** [comment]
CC: @fpletz @grahamc

File: pkgs/development/haskell-modules/hackage-packages.nix
Comment (peti) [inline pkgs/development/haskell-modules/hackage-packages.nix]: The new version fails its test suite on NixOS: https://nix-cache.s3.amazonaws.com/log/fb2a5pp8l6k1ys5gjvj1djh9r8ba27bk-hocker-1.0.2.drv

Comment (ixmatus) [reply pkgs/development/haskell-modules/hackage-packages.nix]: Thank you for pointing that out, I had a late night oopsie and forgot to include an override that fixes this issue. I also need to take the temporary override I put in place in the test's `machine.nix`.

I also see the `hackage-packages.nix` has been updated so I'll drop this hunk.

Comment (ixmatus) [reply pkgs/development/haskell-modules/hackage-packages.nix]: It's failing its test suite due to a missing data file :-/ I thought I'd fixed that but apparently not. Since we need to override the derivation to wrap the programs with a PATH to nix I added a `dontCheck` for now; when I fix this fully I'll open another PR to remove the `dontCheck`.

File: pkgs/development/haskell-modules/configuration-common.nix
Comment (peti) [inline pkgs/development/haskell-modules/configuration-common.nix]: I guess the package should declare its dependency on Nix, i.e. by depending on http://hackage.haskell.org/package/nix-paths?

Comment (ixmatus) [reply pkgs/development/haskell-modules/configuration-common.nix]: The only utility from the hocker package that depends on Nix (other than the tests) is `docker2nix`; I think that's clear enough for now. I agree that a build-time dependency on Nix is better than a run-time dependency, so thank you for pointing me at `nix-paths`. Here are my reasons why I don't think we should gate this PR on that, though:

- I don't entirely understand how `nix-paths` works; it looks like it has magic constants in it but I don't understand where those come from; a minimal README.md for the project would help a lot I think (note that if you explain it all to me, I'm also happy to submit a PR with a README.md contribution)
- The utility that we need for `docker2nix` is `nix-hash`, which is not provided by `nix-paths`. I can submit a PR to fix that but I don't think we should gate this PR on that and I also want to understand how `nix-paths` works first before contributing more functionality

Comment (peti) [reply pkgs/development/haskell-modules/configuration-common.nix]: http://hackage.haskell.org/package/nix-paths-1.0.1 adds support for `nix-hash`. The paths are determined are `configure`-time by Cabal. Look at `Setup.lhs` for the implementation.

Comment (ixmatus) [reply pkgs/development/haskell-modules/configuration-common.nix]: Thanks for adding support for `nix-hash` @peti. I'll probably get to that soon. In the meantime can I get a review of the other parts of this PR?

I don't think using `nix-paths` is a blocker for this PR but if it is, could you say so?

Comment (peti) [reply pkgs/development/haskell-modules/configuration-common.nix]: I would not call the `hocker`-related overrides a *blocker*. That's too strong a term, IMHO. What I can say though is that *I* won't merge the PR as-is. Other Nixpkgs maintainers may very well feel differently about it.

Comment (ixmatus) [reply pkgs/development/haskell-modules/configuration-common.nix]: Okay, thanks for clarifying!

I'd like to clarify my understanding of `nix-paths` (would accept a PR adding a README.md with anything I learn? That may help newcomers and provide a reference to point people at).

I see now that `Setup.lhs` discovers the absolute path to, say, `nix-hash` using `lookupProgram` which is great because if `nix-hash` is not available, it becomes a compile-time error and not a run-time error. Doing this removes the need for the `wrapProgram` invocations but it does require that `nix-paths` is included in the `buildInputs` of `hocker`'s derivation (as a build-time _and_ run-time dependency). Do I have this right?

Additionally, I think the generated derivation for the `nix-paths` hackage package only includes the Nix package in the `nativeBuildInputs`; shouldn't it be in the `propagatedBuildInputs` so that the same Nix used as a build-time dependency is included in the run-time closure of any program that depends on `nix-paths`?

Also, the generated derivation for the `nix-paths` hackage package restricts the hydra build to Linux: `hydraPlatforms = [ "i686-linux" "x86_64-linux" ];` what's the reason for this? `hocker` is usable on Darwin systems, will the `hydraPlatforms` line prevent `hocker` from being integrated on a Darwin worker?

**Comment by ixmatus** [comment]
@peti `hocker-1.0.4` addresses your concerns (it uses `nix-paths`). As part of merging `master` into my branch and resolving the conflict in `configuration-common.nix` I removed the `hocker` override that wrapped the executables with the output from the `nix` package.

**Review by Gabriella439 — APPROVED** [review]

**Comment by ixmatus** [comment]
@peti this is ready for re-review. Perhaps @grahamc would like to look at this too since he was interested in this work.

File: pkgs/development/haskell-modules/configuration-common.nix
Comment (peti) [inline pkgs/development/haskell-modules/configuration-common.nix]: That change doesn't seem useful. :-)

**Review by peti — APPROVED** [review]

**Comment by ixmatus** [comment]
Note that the cabal tests for `hocker` are failing because I didn't correctly include the tests data dir in `hocker`'s cabal file. I have this fixed and am releasing to Hackage soon.

I discovered this by building the nixos test myself off of this branch which means I think the nixos test on this PR isn't included in the checks run by CI.

@grahamc is there anyway to mark small or light-weight nixos tests on a PR as required for merge?

**Comment by ixmatus** [comment]
The fixes for the tests have been [uploaded to hackage](https://hackage.haskell.org/package/hocker-1.0.5) so the next time the hackage packages are updated the nixos test in this PR should be working (I will double check it.)

**Comment by ixmatus** [comment]
I've run the included nixos test and verified it works.

**Comment by adelbertc** [comment]
Ping 👍 am interested in using this as well

**Comment by fpletz** [comment]
@GrahamcOfBorg test hocker-fetchdocker

**Comment by fpletz** [comment]
Not sure why @GrahamcOfBorg is failing but this works for me. :+1:

cc @grahamc

**Review by fpletz — APPROVED** [review]

**Comment by ixmatus** [comment]
Thank you @fpletz!

**Comment by grahamc** [comment]
Is this possibly a Nix 2 vs. Nix 1 problem?

**Comment by ixmatus** [comment]
@grahamc I'm not sure, I will test it tomorrow morning using Nix 2 to be sure.

**Pull request merged**
